dace/frontend/python/newast.py

```diff
@@ -1970,46 +1971,116 @@ def visit_For(self, node: ast.For):
+        self.loop_idx += 1
+        self.continue_states.append([])
+        self.break_states.append([])
         laststate, first_loop_state, last_loop_state = \
             self._recursive_visit(node.body, 'while', node.lineno)
         end_loop_state = self.last_state
 
         # Add loop to SDFG
         loop_cond = astutils.unparse(node.test)
-        self.sdfg.add_loop(laststate, first_loop_state, end_loop_state, None,
-                           None, loop_cond, None, last_loop_state)
+        _, loop_guard, loop_end = self.sdfg.add_loop(
+            laststate, first_loop_state, end_loop_state, None,
+            None, loop_cond, None, last_loop_state)
+        
+        continue_states = self.continue_states.pop()
+        while continue_states:
+            next_state = continue_states.pop()
+            out_edges = self.sdfg.out_edges(next_state)
+            for e in out_edges:
+                self.sdfg.remove_edge(e)
+            self.sdfg.add_edge(next_state, loop_guard, dace.InterstateEdge())
+        break_states = self.break_states.pop()
+        while break_states:
+            next_state = break_states.pop()
+            out_edges = self.sdfg.out_edges(next_state)
+            for e in out_edges:
+                self.sdfg.remove_edge(e)
+            self.sdfg.add_edge(next_state, loop_end, dace.InterstateEdge())
+        self.loop_idx -= 1
+
+    def visit_Break(self, node: ast.Break):
+        if self.loop_idx < 0:
+            error_msg = "'break' is only supported inside for and while loops "
+            if self.nested:
+                error_msg += ("('break' is not supported in Maps and cannot be "
+                              " used in nested DaCe program calls to break out "
+                              " of loops of outer scopes)")
+            raise DaceSyntaxError(self, node, error_msg)
+        self.break_states[self.loop_idx].append(self.last_state)
+    
+    def visit_Continue(self, node: ast.Continue):
+        if self.loop_idx < 0:
+            error_msg = ("'continue' is only supported inside for and while "
+                         "loops ")
+            if self.nested:
+                error_msg += ("('continue' is not supported in Maps and cannot "
+                              " be used in nested DaCe program calls to "
+                              " continue loops of outer scopes)")
+            raise DaceSyntaxError(self, node, error_msg)
+        self.continue_states[self.loop_idx].append(self.last_state)
 
     def visit_If(self, node: ast.If):
         # Add a guard state
         self._add_state('if_guard')
+        if (isinstance(node.test, ast.Compare) and
+                isinstance(node.test.left, ast.Subscript)):
+            cond = self.visit(node.test)
+            if cond in self.sdfg.arrays:
+                cond_dt = self.sdfg.arrays[cond]
+                if isinstance(cond_dt, data.Array):
+                    cond += '[0]'
+            cond_else = 'not ({})'.format(cond)
```

Comment:
The above will be executed only when the if-test is a comparison of the form `data < sth`, i.e. there is a data-dependency. This will generate a boolean scalar or a boolean array of size 1, which in turn will be used in the interstate edge condition. Therefore, 'not ({})'.format(cond)` should always work. I thought that `negate_expr` works only with symbolic expressions, that's why I didn't use it.